### ISSUE TITLE
deprecated plugin middleware is replaced by what ?

### ISSUE THREAD

**Issue by marco-m** [technomancy/leiningen#2387]
Hello,
plugin middleware is indicated as [deprecated since 2.8.0](https://github.com/technomancy/leiningen/blob/master/doc/PLUGINS.md#project-middleware).
What will be its replacement ? What a new plugin should use instead ?
See for example https://github.com/RickMoynihan/lein-tools-deps/issues/3

**Comment by sunng87** [comment]
I'm curious about the replacement.

I have two plugins both of them rely on middleware mechanism to function. To be honest I think middleware design is really elegant. It makes leiningen extensible at any point.

**Comment by technomancy** [comment]
I'm thinking about this more, and I'm convinced that the way hooks and middleware work right now is very wrong. But I think we can salvage the good parts. Hooks are altogether not worth supporting and should be dropped entirely. (Because of how hooks work, we can't stop them entirely from being used; any plugin could add hooks. But we shouldn't have explicit support for them in Leiningen, because that makes it sound like they are endorsed.) But the main problem with middleware is that it's applied automatically without additional configuration.

So I would maybe rephrase it to say that hooks and implicit middleware are deprecated. Middleware is a powerful feature, so if you want to use something that drastic, adding an additional line of config to activate it doesn't seem like too much to ask.

**Comment by sunng87** [comment]
It makes sense. Middleware is just enhanced profiles, with computing functionality. Explicit config like profile works for me.

**Comment by bbatsov** [comment]
@technomancy In cider-nrepl (and similar projects) we use Lein middleware to rewrite the project settings and add there deps, REPL options, etc. See https://github.com/clojure-emacs/cider-nrepl/blob/master/src/cider_nrepl/plugin.clj#L22 as one example.

What are we supposed to do now? As those plugins are supposed to be auto-injected in Lein from the command line we can't really tell users to just modify their project configs.

**Comment by technomancy** [comment]
Middleware is not deprecated; only the implicit loading of middleware thru magic detection in the jar. I will update the title of the bug to be clearer.

Automatic changes to the project map have been shown to cause a lot of breakage that is difficult to debug, so in the future we want to move towards making these things declarative instead of mysterious.

**Comment by bbatsov** [comment]
> Middleware is not deprecated; only the implicit loading of middleware thru magic detection in the jar. I will update the title of the bug to be clearer.

Is there a way to specify this explicitly via the command-line? I also saw how you can make middleware explicit in a project profile and that doesn't work for CIDER's jack-in.

**Comment by technomancy** [comment]
Thinking about this more, I am starting to think that I will remove the deprecation warning.

I would like to deprecate it, but I don't know for sure if there will ever be a Leiningen 3.0, so it doesn't make sense to warn about having it break if it's likely never to happen.

**Comment by bbatsov** [comment]
Thanks for the update! Do you plan to issue 2.8.4 any time soon?

**Comment by didibus** [comment]
Why not just make it a command line arg? Anything that starts lein on its own could decide to inject middlewares that auto-activate, or pass to lein an arg that tells it explicitly what middleware to activate, that seems to be good UX from a user perspective of such tools, like Cider for example.
